### DISCUSSION TITLE
Unable to install certain systemd reliant packages/start services

### DISCUSSION THREAD

**Discussion by blakeduffey** [microsoft/WSL#9350] [General] [answered]
I'm testing installing some packages and I run into a number that (I believe) rely on systemd.  during the setup I get a variety of failures (some listed below, most are 'transport endpoint not detected'

It downloads the packages but once it gets to the install/configuration part at 74% it slows to a crawl and I start to see errors:

Setting up acpi-support (0.144) ...
Failed to restart acpid.service: Transport endpoint is not connected
See system logs and 'systemctl status acpid.service' for details.
Failed to get properties: Transport endpoint is not connected

Setting up openvpn (2.5.5-1ubuntu3.1) ...
Failed to reload daemon: Transport endpoint is not connected
Failed to preset unit: Transport endpoint is not connected
/usr/bin/deb-systemd-helper: error: systemctl preset failed on openvpn.service: No such file or directory

Setting up ubuntu-drivers-common (1:0.9.6.1) ...
Failed to preset unit: Transport endpoint is not connected
/usr/bin/deb-systemd-helper: error: systemctl preset failed on gpu-manager.service: No such file or directory

Setting up whoopsie (0.2.77) ...
Failed to preset unit: Transport endpoint is not connected
/usr/bin/deb-systemd-helper: error: systemctl preset failed on whoopsie.path: No such file or directory
Failed to reload daemon: Transport endpoint is not connected

Setting up power-profiles-daemon (0.10.1-3) ...
Failed to preset unit: Transport endpoint is not connected
/usr/bin/deb-systemd-helper: error: systemctl preset failed on power-profiles-daemon.service: No such file or directory
Failed to reload daemon: Transport endpoint is not connected
Failed to get unit file state for power-profiles-daemon.service: Transport endpoint is not connected
Failed to retrieve unit state: Transport endpoint is not connected
power-profiles-daemon.service is a disabled or a static unit, not starting it.

Setting up upower (0.99.17-1) ...
Failed to reload daemon: Transport endpoint is not connected
Failed to get unit file state for upower.service: Transport endpoint is not connected
Failed to retrieve unit state: Transport endpoint is not connected
upower.service is a disabled or a static unit, not starting it.

Setting up secureboot-db (1.8) ...
Failed to preset unit: Transport endpoint is not connected
/usr/bin/deb-systemd-helper: error: systemctl preset failed on secureboot-db.service: No such file or directory
Failed to reload daemon: Transport endpoint is not connected

(this is just a sample)

**Comment by pjmattingly** [comment]
Also getting this error when:

```
sudo apt install ubuntu-desktop gnome
```

**Comment by childishtree** [comment]
I report the same error using the  `sudo apt install ubuntu-desktop`  command.

**Reply by robin-collins** [reply]
same

**Reply by Biswa96** [reply]
This seems to be an old issue. Have you tried that after enabling systemd in WSL2?

**Reply by BroSupermarketMe** [reply]
I have tried enabling systemd in WSL2, still same

**Comment by iCodeDesigns** [comment]
Same issue sudo apt install ubuntu-desktop

**Comment by prunckle** [comment]
I'm getting a similar issue when attempting to query systemd status via systemctl or when trying to restart failed services via systemctl:

`╭─username at rhel8base (WSL: rhel8-devops | kernel: 5.15.90.1-microsoft-standard-WSL2) in /mnt/c/Users/UserName`
`╰─ 23-04-28 - 12:40:26 ○ sudo systemctl status`
`Failed to read server status: Transport endpoint is not connected`

and

`╭─username at rhel8base (WSL: rhel8-devops | kernel: 5.15.90.1-microsoft-standard-WSL2) in /mnt/c/Users/UserName`
`╰─ 23-04-28 - 12:41:42 ○ sudo systemctl restart dnf-makecache.service`
`Failed to restart dnf-makecache.service: Transport endpoint is not connected`
`See system logs and 'systemctl status dnf-makecache.service' for details.`

**Comment by blakeduffey** [comment]
this persists as of WSL 1.2.5.0

I tried to install gnome (again) and got hung at the exact same point

**Comment by gf9276** [comment]
`sudo apt remove --purge acpi-support` can solve this problem. But this still prevents the installation of gnome UI.

**Comment by KamenRevice** [comment]
same issue with wsl2 and ` sudo  apt install ubuntu-desktop gnome `  , how do you  solve this problem

**Comment by raarts** [comment, marked as answer]
Use this before installing ubuntu-desktop:

```
sudo apt-mark hold acpid acpi-support
```

(The `hold` option prevents a package from being automatically installed, upgraded or removed.)

**Reply by blakeduffey** [reply]
> ```
> sudo apt-mark hold acpid acpi-support
> ```

I am able to install ubuntu-desktop using this method.  

I cannot connect to the desktop however, after installing it.

**Reply by Esoteric-Coder** [reply]
Does create any problem afterwards ?
Can I still upgrade my packages using ` upgrade` command ?

**Reply by p711** [reply]
@Esoteric-Coder No problems, upgrade still possible.

**Reply by nucleare** [reply]
I couldn't run the hold command because running `apt` in any form (i.e. `apt-mark`) would result in telling me about how `dpkg` was interrupted* since I had tried to get out of the hang by hitting `Ctrl`+`C` So for anyone else finding themselves after the fact, it wasn't until I ran  `sudo rm -rf /etc/acpi/support` did it free up `apt` to run the `...hold acpid acpi-support` command. Then I just ran `dpkg --configure --pending` afterwards and the rest of the install went smoothly.

_*Error message referenced:_ `E: dpkg was interrupted, you must manually run 'sudo dpkg --configure -a' to correct the problem.`

**Comment by raarts** [comment]
@blakeduffey After installing ubuntu-desktop you should do the following:

```
sudo apt install xrdp
sudo sed -i 's/3389/3390/g' /etc/xrdp/xrdp.ini    # change the Linux RDP port, because already used by Windows 
sudo systemctl restart xrdp
```

Next create the file `.xsessionrc` in your Linux home directory with the following contents:

```
export GNOME_SHELL_SESSION_MODE=ubuntu
export XDG_CURRENT_DESKTOP=ubuntu:GNOME
export XDG_DATA_DIRS=/usr/share/ubuntu:/usr/local/share:/usr/share:/var/lib/snapd/desktop
export WAYLAND_DISPLAY=
export XDG_CONFIG_DIRS=/etc/xdg/xdg-ubuntu:/etc/xdg
```

After that connect to localhost:3390 using Remote Desktop.

**Reply by gf9276** [reply]
Hello, I would like to ask if there may be a slight lag when connecting to the remote desktop using xrdp. Wsl2 cannot use a virtual graphics card for image acceleration in this situation

**Reply by gf9276** [reply]
After my trial, using Ubuntu 22.04+gnome4 can provide the best FPS experience

**Reply by blakeduffey** [reply]
@raarts - OUTSTANDING!  I'm doing it again to verify my install steps - but this looks like a winner!

**Reply by hrishikesh-k** [reply]
Thanks @blakeduffey. I was able to follow your instructions and log in to xrdp. However, I see not much than a black screen like this:

![image](https://github.com/microsoft/WSL/assets/11403551/b8fd619b-e4fa-41bb-93f3-583d0e9b65a9)

Can't use this at all. I tried installing gnome as well, but didn't change much.

The mouse is moving smoothly, so I assume it's working fine. But apps are hardly usable like this.

**Reply by blakeduffey** [reply]
That almost looks like a refresh rate issue...

**Reply by hrishikesh-k** [reply]
Thanks. Would you happen to know how to resolve it?

**Reply by Heng-Zhou** [reply]
Don't forget to reboot before first connection.

**Comment by medaminezghal** [comment]
apt purge -y acpid acpi-support modemmanager
apt-mark hold acpid acpi-support modemmanager
then reboot

**Reply by luchiayu** [reply]
this comment should put on top of this discussion. Very important.
CY

**Reply by chaodonghe** [reply]
I got the same problem. Is it working for you by doing:
---
apt purge -y acpid acpi-support modemmanager
apt-mark hold acpid acpi-support modemmanager
then reboot
---

**Comment by Zephyrel** [comment]
I've been just getting into WSL2 and been trying to get Ubuntu Linux running with GNOME and I've been all over the place withtrying to get the desktop environment to "Just Work" stably with vcXsrv without crashing, or freezing and bloating with memory and cpu usage until it stopped running around 74% of a 'sudo apt install ubuntu-desktop', same issues as OP. I can say that I've got it working with much trial and error (for example, running just gnome-session at the end would just open the desktop environment over my entire dual monitor setup, with some methods letting it run albeit unstable and poorly, or not at all with a error screen on start). Although there are some quirks left to be ironed out (later, I got other things to do now that I've gotten to a great breakpoint). I figure this can get a few people having troubles started.

First off, some specs.
```
Windows 10 Education 22H2
Intel i7-7700k, ASUS z270-A  
# wsl --version
WSL version: 1.2.5.0
Kernel version: 5.15.90.1
WSLg version: 1.0.51
MSRDC version: 1.2.3770
Direct3D version: 1.608.2-61064218
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.3636
```

Things I've learned:

1. systemd is quirky (at least on Windows 10, don't know about Windows 11)
2. acpi-support just outright bricks the entire distro.

Points of reference.
        https://github.com/microsoft/wslg/tree/main
	https://devblogs.microsoft.com/commandline/systemd-support-is-now-available-in-wsl/
	https://askubuntu.com/questions/1442663/cannot-start-gnome-session-after-enabling-systemd-on-latest-wsl2
	https://x410.dev/cookbook/wsl/using-x410-with-wsl2/
	https://gist.github.com/Ta180m/e1471413f62e3ed94e72001d42e77e22
	https://github.com/DamionGans/ubuntu-wsl2-systemd-script
	https://github.com/microsoft/WSL/discussions/9350
	https://github.com/microsoft/WSL/issues/10498
	https://github.com/microsoft/WSL/issues/10356
	https://github.com/microsoft/WSL/issues/10059
	https://github.com/microsoft/WSL/issues/9029
	https://github.com/microsoft/WSL/issues/9013
	https://github.com/microsoft/WSL/issues/8883
	https://github.com/microsoft/WSL/issues/8867
	https://github.com/microsoft/WSL/issues/8904
	https://github.com/microsoft/WSL/issues/5816

The following quirks and features for setup that need to be looked at after this and written.

0. Don't run gnome-session. Start up command at the end is going to be:
`MUTTER_DEBUG_DUMMY_MODE_SPECS=1920x1020 gnome-shell --nested --no-x11 2> /dev/null &`
  It generates the following error message while the Desktop environment comes up.
`portal is not running: GDBus.Error:org.freedesktop.DBus.Error.Spawn.ChildExited: Process org.freedesktop.portal.IBus exited with status 1`

1. "Settings" and "Files" are a bit broken. They will open, and you can use them,
    but once you close the windows, you won't be able to reopen.
    You can still open "Files" in a new window with a right click and explore.
2. Opening "System Monitor" or "Thunderbird" works. When closing either window
   it takes down the entire display server, generating the following error message.
   `Bail out! libmutter:ERROR:../src/core/window.c:1544:meta_window_unmanage: assertion failed: (window->display->focus_window != window)`
   You can ctrl+C in your WSL Linux shell and restart the desktop environment afterwards with the same command (you'll see later)
   'systemctl status' will also return your state as 'running' and not 'degraded' so that's a plus
3. I need to test out/get pulseaudio working such that sound can play from the Ubuntu Desktop.
   Supposedly WSLg can handle this. I'll see about this later.
   https://www.freedesktop.org/wiki/Software/PulseAudio/Ports/Windows/Support/
4. I've only tested and gotten gnome working. Other desktop environments remain to be seen.
   I would like to get budgie working, personally.
5. I'm sure there's more quirks that will rear their heads later.
    I'm not going to dive immediately back into the nitty gritty just yet to find out how to fix or work around more issues.

Start with vcxsrv. You can install the latest available version with chocolatey
or from https://sourceforge.net/projects/vcxsrv/
Default install location is C:\Program Files\VcXsrv\xlaunch.exe
Run it, and use the following install options:
Multiple Windows
Display Number: 0
Start no client
Check all options: Clipboard, Primary Selection, Native opengl, Disable access control

From a fresh install/start, follow along and do the following:
sudo apt update
sudo apt upgrade
//Enable systemd in either one of these two ways. Try 1st one first.
//Method 1) Edit the wsl.conf file.
	sudo vim /etc/wsl.conf
	Add these lines then save
		[boot]
		systemd=true
//Method 2 (to be tested)  Still in WSL2 Ubuntu linux
	git clone https://github.com/DamionGans/ubuntu-wsl2-systemd-script.git
	cd ubuntu-wsl2-systemd-script/
	bash ubuntu-wsl2-systemd-script.sh
//create directories and file
sudo mkdir /etc/pam.d/system
sudo vim /etc/pam.d/system/system-user
```
	# This file is part of systemd.
	#
	# Used by systemd --user instances.
	
	@include common-account

	session  required pam_loginuid.so
	session  required pam_limits.so
	@include common-session-noninteractive
	session optional pam_systemd.so
```
sudo mkdir /run/user/1000
sudo chmod 700 /run/user/1000
sudo chown $(whoami): /run/user/1000
Shutdown from second terminal 'wsl --shutdown'
Start it up again. 'wsl -d ubuntu'
//DO NOT LEAVE OUT THE '-' after acpi-support or you WILL brick the distro,
//and then you'll have to --terminate, --unregister with wsl and start all over
sudo apt install ubuntu-desktop acpi-support-
	//to test, for myself, later
	//'sudo apt install ubuntu-budgie-desktop acpi-support-' (choose lightdm when prompted) (I like budgie)
//Just in case. Thanks to: https://github.com/medaminezghal
sudo apt purge -y acpid acpi-support modemmanager
sudo apt-mark hold acpid acpi-support modemmanager
//The following comments are irrelevant if you just don't install the acpi-support to begin with
//but in the case that was installed
	//disable acpid, acpi-support just breaks the distro
	//sudo ln -s /dev/null /etc/systemd/system/acpid.service
	//sudo ln -s /dev/null /etc/systemd/system/acpid.path
	//sudo apt purge -y acpid acpi-support modemmanager
	//sudo apt-mark hold acpid acpi-support modemmanager
//this is for testing gui later, in case this is not installed. Getting xeyes.
sudo apt-get install x11-apps

'wsl --shutdown' from terminal 2, reboot in term 1

sudo vim ~/.bashrc, paste and save the following:
	//This magic line makes your display server work
	export DISPLAY=$(cat /etc/resolv.conf | grep nameserver | awk '{print $2}'):0
	//This magic line is for vcxsrv
	export LIBGL_ALWAYS_INDIRECT=1
	//
	XDG_SESSION_TYPE='x11 gnome-session --session=ubuntu --systemd-service > /dev/null 2>&1'

'wsl --shutdown' from terminal 2, 'wsl -d ubuntu' in term 1

//this should pull up a big table of services if everything went off correctly. q to quit.
systemctl list-unit-files --type=service
//now run
xeyes
//you should see a window with some mouse-tracking googly eyes.
//right click on vcxsrv in your taskbar, applications > xwininfo. It will open a terminal.
//click on the xeyes window. the xwininfo terminal should populate with information.
//you can close both xeyes and xwininfo now

//don't run gnome-session, it doesn't/won't work.
//run gnome-shell just like this, change the resolution as you please
MUTTER_DEBUG_DUMMY_MODE_SPECS=1920x1020 gnome-shell --nested --no-x11 2> /dev/null &
//open up firefox and be glad that that "just works". You can also pull it out of the ubuntu window and tile it within Windows.

//And that's all I've got so far.
![image_2023-11-06_153943038](https://github.com/microsoft/WSL/assets/143488604/6d036861-bbc0-4f77-9cc7-2458c58b5573)

Any additions are greatly welcome.

**Comment by blakeduffey** [comment]
ACPI definitely bricks the install.

Here are my steps (tested several times on Ubuntu 22.04).  This uses xrdp vs vcXsrv.  I find that approach cleaner.

```
sudo apt-mark hold acpid acpi-support
sudo apt update && sudo apt upgrade -y
sudo apt -y install xrdp
sudo cp /etc/xrdp/xrdp.ini /etc/xrdp/xrdp.ini.bak
sudo sed -i 's/3389/3390/g' /etc/xrdp/xrdp.ini
sudo sed -i 's/max_bpp=32/#max_bpp=32\nmax_bpp=128/g' /etc/xrdp/xrdp.ini
sudo sed -i 's/xserverbpp=24/#xserverbpp=24\nxserverbpp=128/g' /etc/xrdp/xrdp.ini
```


Then create ~/.xsessionrc with the following contents:

```
export GNOME_SHELL_SESSION_MODE=ubuntu
export XDG_CURRENT_DESKTOP=ubuntu:GNOME
export XDG_DATA_DIRS=/usr/share/ubuntu:/usr/local/share:/usr/share:/var/lib/snapd/desktop
export WAYLAND_DISPLAY=
export XDG_CONFIG_DIRS=/etc/xdg/xdg-ubuntu:/etc/xdg
```

Finally, install the GNOME deskop and xrdp

```
sudo apt install -y ubuntu-desktop-minimal # or ubuntu-desktop
sudo systemctl restart xrdp
```

Connect via the standard Windows mstsc client to localhost:3390.  This will provide a pretty good desktop environment, including multi-monitor support.

FWIW - I have different variations on this set of docs that work for Budgie, Cinnamon, MATE, Xfce, Enlightenment, KDE, LXDE and LXQT. 

GNOME and KDE tend to be the most needy in terms of configuration.  LXDE and LXQT require the least effort.

**Reply by Zephyrel** [reply]
Greatly appreciated. I may ask you if you could gist your setup (especially that budgie setup) for those that come across this. I did see that xrpd solution from raarts up above, just hadn't quite gotten around to trying that yet, I'm not as keen on understanding exactly which tools I can use to achieve great results and how everything works and fits together, so I had been stumbling through the multiple attempted solutions across the web as a piecemeal puzzle.

I'll give this a shot with a fresh install, after I go and pick up the wsl pre-release update.
Cheers!

**Comment by blakeduffey** [comment]
Budgie, Cinnamon, Mate, Xfce are all basically the same (save a change in startwm.sh)

```
sudo apt-mark hold acpid acpi-support
sudo apt update && sudo apt upgrade -y
sudo apt -y install xrdp
sudo cp /etc/xrdp/xrdp.ini /etc/xrdp/xrdp.ini.bak
sudo sed -i 's/3389/3390/g' /etc/xrdp/xrdp.ini
sudo sed -i 's/max_bpp=32/#max_bpp=32\nmax_bpp=128/g' /etc/xrdp/xrdp.ini
sudo sed -i 's/xserverbpp=24/#xserverbpp=24\nxserverbpp=128/g' /etc/xrdp/xrdp.ini
```

Then edit **/etc/xrdp/startwm.sh**

```
unset DBUS_SESSION_BUS_ADDRESS
unset XDG_RUNTIME_DIR
 
#test -x /etc/X11/Xsession && exec /etc/X11/Xsession
#exec /bin/sh /etc/X11/Xsession
 
#uncomment the correct DE below
#budgie-desktop
#cinnamon-session
#mate-session
#startxfce4
```

Then install the desktop and bounce xrdp

```
sudo apt install -y ubuntu-budgie-desktop
sudo systemctl restart xrdp
```

**Comment by vinberg88** [comment]
Nice work - Can't get it to work with 22.04 from Microsoft store. It starts but shuts down just as quickly. 

What am I doing wrong ? 

21.10 and lower version of ubuntu works and i use

https://gist.github.com/Ta180m/e1471413f62e3ed94e72001d42e77e22

Regards,
Vinberg

**Comment by blakeduffey** [comment]
Fwiw all my testing is done with 22.04

**Reply by vinberg88** [reply]
OK - Strange

**Comment by vinberg88** [comment]
Gnome - Easy setup

...................................
Download and installl Ubuntu 21.04

https://cloud-images.ubuntu.com/releases/hirsute/release/ubuntu-21.04-server-cloudimg-amd64-wsl.rootfs.tar.gz

Follow This and Paste in terminal - Ubuntu - No stress... =)
------------------
sudo nano /etc/wsl.conf 

and add

[boot]
systemd=true

-------------------
Next
-------------------

ln -s libcuda.so.1.1 libcuda.so.1
ln -s libcuda.so.1.1 libcuda.so
echo -e "[automount]\nldconfig = false" | sudo tee -a /etc/wsl.conf
sudo mkdir /usr/lib/wsl/lib2
sudo ln -s /usr/lib/wsl/lib/* /usr/lib/wsl/lib2
echo /usr/lib/wsl/lib2 | sudo tee /etc/ld.so.conf.d/ld.wsl.conf

-------------------
Next
-------------------

sudo ln -s /dev/null /etc/systemd/system/acpid.service
sudo ln -s /dev/null /etc/systemd/system/acpid.path

-------------------
Next
-------------------

sudo cp /etc/apt/sources.list /etc/apt/sources.list.bak
sudo sed -i -re 's/([a-z]{2}.)?archive.ubuntu.com|security.ubuntu.com/old-releases.ubuntu.com/g' /etc/apt/sources.list
sudo apt-get update && sudo apt-get dist-upgrade

-------------------
Next
-------------------

sudo apt install bash git npm wget curl nano sudo pipx dracut firefox -y
sudo apt install gdm3* -y
sudo systemctl set-default graphical

-------------------
Next
-------------------

sudo pipx install hyfetch 
pipx ensurepath

-------------------
Next
-------------------

Now install and config GWSL

GWSL can be find here - https://apps.microsoft.com/detail/gwsl/9NL6KD1H33V3?hl=en-US&gl=US

-------------------
Next
-------------------

git clone https://github.com/DamionGans/ubuntu-wsl2-systemd-script.git
cd ubuntu-wsl2-systemd-script/
bash ubuntu-wsl2-systemd-script.sh

And reboot Ubuntu

-------------------
Next
-------------------

sudo apt install ubuntu-desktop gnome

And reboot Ubuntu

-------------------
Next
-------------------

Start Gnome - type in terminal =)

gnome-session

-------------------

Done and Enjoy 

https://i.imgur.com/O8KBqkc.png 

-------------------

**Reply by ravensorrow** [reply]
Would this work with a WSL2/Debian-Testing install?

**Comment by vinberg88** [comment]
Done, hard work =)  Ubuntu 22.04 and Windows 11 Pro

https://i.imgur.com/RResBn4.png

Took me two days and of course I found another way to solve this big problem.

**Comment by vinberg88** [comment]
Ubuntu 23.10 - Gnome - Windows 11 Pro!

https://i.imgur.com/HCJrAlc.png

I will celebrate this with a whiskey!  - I am on fire!

**Reply by TharaBhaiDoraemon** [reply]
get this man a whiskey!

**Comment by gitanoqevaporelmundoentero** [comment]
I keep having `Transport endpoint is not connected` errors on a fresh new WSL2 install of Ubuntu 24.04. I also notice that `acpi-support` doesn't seem to exist on Ubuntu 24.04:

```
$ sudo apt-mark hold acpid acpi-support
acpid was already set on hold.
E: Unable to locate package acpi-support
```

**Reply by gitanoqevaporelmundoentero** [reply]
I've just discovered that `Transport endpoint is not connected` errors seem to disappear when stopping `Docker Desktop` in Windows, which is integrated with WSL2.

**Comment by zbaibg** [comment]
@blakeduffey 's [answer on Gnome](https://github.com/microsoft/WSL/discussions/9350#discussioncomment-7492570) really works for me. Thanks!
I have some notes on @blakeduffey 's [answer about other Desktop Envirenments other than Gnome](https://github.com/microsoft/WSL/discussions/9350#discussioncomment-7492749) :

If you want to use xfce4 etc. other than GNOME, you still need to turn **wslg** off by some way.

Any of the following ways should works:

1. You could turn it off by setting **guiApplications** variable in **.wslconfig** file ([see here](https://learn.microsoft.com/en-us/windows/wsl/wsl-config))

2. You could add `unset WAYLAND_DISPLAY`at the head of ** /etc/xrdp/startwm.sh** and modify it as[ this](https://github.com/microsoft/WSL/discussions/9350#discussioncomment-7492749).
  (tip1: For original **startwm.sh** without modification, **~/.xsessionrc** would be loaded during RDP connection. If you modify **startwm.sh** in this way, **~/.xsessionrc** would not be loaded during your RDP)
  (tip2: If you modify **startwm.sh** in this way, you may need  `export XAUTHORITY=/home/${USER}/.Xauthority ` if you meet some X11 Authentication problem for some applications like firefox)


3. You could just leave  **/etc/xrdp/startwm.sh** as original. Instead you modify **~/.xsessionrc** as the following. (I belive this is more standard)

```
export WAYLAND_DISPLAY=
exec /usr/bin/xfce4-session
```


BTW, these desktop envirenment processes may always leave some processes in the backend. It is a good idea to keep `pkill -u -9 $USER` (to kill all your processes), or `wsl --shutdown` in Windows end to make your processes clean during debuging. 

For some reason, if I use ` unset WAYLAND_DISPLAY` instead of ` export WAYLAND_DISPLAY=`, each time I shutdown **wsl** and reboot it, I cannot RDP to XFCE4/GNOME. It just crashed. I have to `pkill -u -9 $USER` and RDP again, then it works. However some applications like firefox does detect one empty `WAYLAND_DISPLAY` variable. So ` export WAYLAND_DISPLAY=` will not let them work. I do not find workaround for this yet.
